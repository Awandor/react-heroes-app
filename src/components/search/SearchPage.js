import { HeroCard } from '../heroes/HeroCard';
import { useForm } from "../../hooks/useForm";
import { useHistory, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { useMemo } from 'react';

export const SearchPage = () => {

    const history = useHistory();

    const location = useLocation();

    const { q = '' } = queryString.parse( location.search ); // desestructuramos e igualamos a string vacío por defecto para evitar undefined

    const [ { name }, handleInputChange ] = useForm( {
        name: q // tiene que ser el nombre del input y lo igualamos a q así si hemos buscado y refrescado mantenemos el valor de q
    } );

    // const heroesFiltered = getHeroesByName(name);

    const heroesFiltered = useMemo( () => getHeroesByName( q ), [ q ] ); // Evita que se dispare cada vez que toco una tecla, sólo cuando q cambia



    const handleSearch = ( e ) => {

        e.preventDefault(); // evitar el refresco de la página

        // console.log( history );

        // console.log( location ); // nos da la propiedad search

        // console.log( 'queryString', queryString.parse( location.search ) ); // Nos devuelve un objeto con todos los parámetros de búsqueda

        // console.log( 'name +++++++++++++', name );

        if ( name.trim().length <= 1 ) {

            return;

        }

        history.push( `?q=${name}` ); // Nos empuja a la misma página con el parámetro y gracias a queryString.parse cambiamos el valor de q

    };

    return (
        <div>
            <h1>Search Page</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input className="form-control" type="text" name="name" placeholder="Find your hero" autoComplete="off" value={name} onChange={handleInputChange} />

                        <button className="btn btn-primary btn-block mt-3" type="submit">Search</button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {
                        ( q === '' ) && <div className="alert alert-info">
                            Search for a hero
                        </div>
                    }

{
                        ( q !== '' && heroesFiltered.length === 0 ) && <div className="alert alert-danger">
                            There is no hero with name {q}
                        </div>
                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ) )
                    }
                </div>
            </div>
        </div>
    );

};
