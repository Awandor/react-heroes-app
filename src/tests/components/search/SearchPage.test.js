import { mount } from "enzyme";
import { MemoryRouter, Route, Router } from "react-router-dom";
import { SearchPage } from "../../../components/search/SearchPage";


describe( 'Pruebas en <SearchPage />', () => {

    test( 'debe mostrarse correctamente con valor por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search' ]}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find( '.alert-info' ).text().trim() ).toBe( 'Search for a hero' );

    } );

    test( 'debe mostrar a Batman y el input con el valor del queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );

        expect( wrapper.find( 'input' ).prop( 'value' ) ).toBe( 'batman' );

    } );

    test( 'debe mostrar un error si no se encuentra el héroe', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=algo' ]}>
                <Route path="/search" component={SearchPage} />
            </MemoryRouter>
        );

        expect( wrapper.find( '.alert-danger' ).text().trim() ).toBe( 'There is no hero with name algo' );

    } );

    test( 'debe de llamar el push del history', () => {

        const historyMock = {
            push: jest.fn(),

            // Nos fijamos en los errores de lo que falta y los vamos añadiendo deduciendo el tipo
            listen: jest.fn(),
            createHref: jest.fn(),
            replace: jest.fn(),
            location: {
                pathname: '/search' // Debe coincidir con Route
            }
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search' ]}>
                <Router history={historyMock}>
                    <Route path="/search" component={ () => <SearchPage />} />
                </Router>
            </MemoryRouter>
        );

        // Simulamos el cambio en el input y enviamos el evento

        wrapper.find( 'input' ).simulate( 'change', {
            target: {
                name: 'name',
                value: 'batman'
            }
        } );

        wrapper.find( 'form' ).prop( 'onSubmit' )( { preventDefault() { } } );

        expect( historyMock.push ).toHaveBeenCalled();
    
} );




} );
