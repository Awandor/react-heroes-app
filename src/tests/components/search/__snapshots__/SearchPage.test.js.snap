// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe mostrarse correctamente con valor por defecto 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary btn-block mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search for a hero
      </div>
    </div>
  </div>
</div>
`;
